File: examples/php-orders-client/orders_client.php
<?php

declare(strict_types=1);

function env_value(string $key, string $fallback): string
{
    $value = getenv($key);

    return $value === false || $value === '' ? $fallback : $value;
}

function status_code(array $headers): int
{
    if ($headers === []) {
        return 0;
    }

    if (preg_match('/^HTTP\/\S+\s+(\d{3})\b/', $headers[0], $matches) !== 1) {
        return 0;
    }

    return (int) $matches[1];
}

function http_request(string $method, string $url, array $headers, string $body): array
{
    $context = stream_context_create([
        'http' => [
            'method' => $method,
            'header' => implode("\r\n", $headers) . "\r\n",
            'content' => $body,
            'timeout' => 10,
            'ignore_errors' => true,
        ],
    ]);

    $raw = file_get_contents($url, false, $context);
    if ($raw === false) {
        throw new RuntimeException("HTTP request failed before response: {$method} {$url}");
    }

    /** @var array<int, string> $http_response_header */
    return [status_code($http_response_header ?? []), $raw];
}

function request_token(
    string $tokenURL,
    string $clientID,
    string $clientSecret,
    string $audience,
    string $scope
): array {
    $form = http_build_query([
        'grant_type' => 'client_credentials',
        'audience' => $audience,
        'scope' => $scope,
    ], '', '&', PHP_QUERY_RFC3986);

    [$status, $raw] = http_request('POST', $tokenURL, [
        'Authorization: Basic ' . base64_encode($clientID . ':' . $clientSecret),
        'Content-Type: application/x-www-form-urlencoded',
        'Accept: application/json',
    ], $form);

    if ($status !== 200) {
        throw new RuntimeException("token request failed: status={$status} body={$raw}");
    }

    $token = json_decode($raw, true);
    if (!is_array($token)) {
        throw new RuntimeException('token response is not valid JSON: ' . json_last_error_msg());
    }

    if (($token['access_token'] ?? '') === '') {
        throw new RuntimeException('token response is missing access_token');
    }
    if (strcasecmp((string) ($token['token_type'] ?? ''), 'Bearer') !== 0) {
        throw new RuntimeException('token response has unexpected token_type=' . (string) ($token['token_type'] ?? ''));
    }
    if ((int) ($token['expires_in'] ?? 0) <= 0) {
        throw new RuntimeException('token response has invalid expires_in=' . (string) ($token['expires_in'] ?? ''));
    }

    return $token;
}

function create_invoice(string $billingURL, string $accessToken): array
{
    $body = json_encode([
        'order_id' => 'ord_demo_001',
        'amount' => 1500,
    ], JSON_THROW_ON_ERROR);

    [$status, $raw] = http_request('POST', $billingURL, [
        'Authorization: Bearer ' . $accessToken,
        'Content-Type: application/json',
        'Accept: application/json',
    ], $body);

    if ($status < 200 || $status >= 300) {
        throw new RuntimeException("billing request failed: status={$status} body={$raw}");
    }

    return [$status, $raw];
}

$tokenURL = env_value('ORIONIS_TOKEN_URL', 'http://localhost:8080/oauth/token');
$billingURL = env_value('BILLING_URL', 'http://localhost:8081/invoices');
$clientID = env_value('ORIONIS_CLIENT_ID', 'orders-service');
$clientSecret = env_value('ORIONIS_CLIENT_SECRET', 'orders-local-secret-change-me');
$audience = env_value('ORIONIS_AUDIENCE', 'billing-api');
$scope = env_value('ORIONIS_SCOPE', 'billing.invoice.create');

$token = request_token($tokenURL, $clientID, $clientSecret, $audience, $scope);
[$status, $raw] = create_invoice($billingURL, (string) $token['access_token']);

echo "status={$status}\n";
echo rtrim($raw) . "\n";

File: examples/java-orders-client/OrdersClient.java
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Base64;

public final class OrdersClient {
    private static final Duration HTTP_TIMEOUT = Duration.ofSeconds(10);

    private OrdersClient() {
    }

    public static void main(String[] args) throws Exception {
        String tokenURL = getenv("ORIONIS_TOKEN_URL", "http://localhost:8080/oauth/token");
        String billingURL = getenv("BILLING_URL", "http://localhost:8081/invoices");
        String clientID = getenv("ORIONIS_CLIENT_ID", "orders-service");
        String clientSecret = getenv("ORIONIS_CLIENT_SECRET", "orders-local-secret-change-me");
        String audience = getenv("ORIONIS_AUDIENCE", "billing-api");
        String scope = getenv("ORIONIS_SCOPE", "billing.invoice.create");

        HttpClient http = HttpClient.newBuilder()
                .connectTimeout(HTTP_TIMEOUT)
                .build();

        TokenResponse token = requestToken(http, tokenURL, clientID, clientSecret, audience, scope);
        HttpResponse<String> billingResponse = createInvoice(http, billingURL, token.accessToken);

        System.out.printf("status=%d%n%s%n", billingResponse.statusCode(), billingResponse.body());
    }

    private static TokenResponse requestToken(
            HttpClient http,
            String tokenURL,
            String clientID,
            String clientSecret,
            String audience,
            String scope
    ) throws IOException, InterruptedException {
        String form = "grant_type=client_credentials"
                + "&audience=" + formValue(audience)
                + "&scope=" + formValue(scope);

        String basicCredentials = Base64.getEncoder()
                .encodeToString((clientID + ":" + clientSecret).getBytes(StandardCharsets.UTF_8));

        HttpRequest request = HttpRequest.newBuilder(URI.create(tokenURL))
                .timeout(HTTP_TIMEOUT)
                .header("Authorization", "Basic " + basicCredentials)
                .header("Content-Type", "application/x-www-form-urlencoded")
                .header("Accept", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(form, StandardCharsets.UTF_8))
                .build();

        HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        if (response.statusCode() != 200) {
            throw new IllegalStateException(
                    "token request failed: status=" + response.statusCode() + " body=" + response.body()
            );
        }

        TokenResponse token = new TokenResponse(
                jsonString(response.body(), "access_token"),
                jsonString(response.body(), "token_type"),
                jsonLong(response.body(), "expires_in"),
                jsonString(response.body(), "scope")
        );

        if (token.accessToken.isBlank()) {
            throw new IllegalStateException("token response is missing access_token");
        }
        if (!"Bearer".equalsIgnoreCase(token.tokenType)) {
            throw new IllegalStateException("token response has unexpected token_type=" + token.tokenType);
        }
        if (token.expiresIn <= 0) {
            throw new IllegalStateException("token response has invalid expires_in=" + token.expiresIn);
        }

        return token;
    }

    private static HttpResponse<String> createInvoice(HttpClient http, String billingURL, String accessToken)
            throws IOException, InterruptedException {
        String body = "{\"order_id\":\"ord_demo_001\",\"amount\":1500}";
        HttpRequest request = HttpRequest.newBuilder(URI.create(billingURL))
                .timeout(HTTP_TIMEOUT)
                .header("Authorization", "Bearer " + accessToken)
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                .build();

        HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IllegalStateException(
                    "billing request failed: status=" + response.statusCode() + " body=" + response.body()
            );
        }

        return response;
    }

    private static String getenv(String key, String fallback) {
        String value = System.getenv(key);
        return value == null || value.isBlank() ? fallback : value;
    }

    private static String formValue(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private static String jsonString(String json, String field) {
        int valueStart = jsonValueStart(json, field);
        if (valueStart < 0 || valueStart >= json.length() || json.charAt(valueStart) != '"') {
            return "";
        }

        StringBuilder value = new StringBuilder();
        boolean escaping = false;
        for (int i = valueStart + 1; i < json.length(); i++) {
            char c = json.charAt(i);
            if (escaping) {
                value.append(unescape(c));
                escaping = false;
                continue;
            }
            if (c == '\\') {
                escaping = true;
                continue;
            }
            if (c == '"') {
                return value.toString();
            }
            value.append(c);
        }

        return "";
    }

    private static long jsonLong(String json, String field) {
        int valueStart = jsonValueStart(json, field);
        if (valueStart < 0) {
            return 0;
        }

        int valueEnd = valueStart;
        while (valueEnd < json.length() && Character.isDigit(json.charAt(valueEnd))) {
            valueEnd++;
        }
        if (valueEnd == valueStart) {
            return 0;
        }

        return Long.parseLong(json.substring(valueStart, valueEnd));
    }

    private static int jsonValueStart(String json, String field) {
        String quotedField = "\"" + field + "\"";
        int fieldIndex = json.indexOf(quotedField);
        if (fieldIndex < 0) {
            return -1;
        }

        int colon = json.indexOf(':', fieldIndex + quotedField.length());
        if (colon < 0) {
            return -1;
        }

        int valueStart = colon + 1;
        while (valueStart < json.length() && Character.isWhitespace(json.charAt(valueStart))) {
            valueStart++;
        }

        return valueStart;
    }

    private static char unescape(char c) {
        switch (c) {
            case '"':
            case '\\':
            case '/':
                return c;
            case 'b':
                return '\b';
            case 'f':
                return '\f';
            case 'n':
                return '\n';
            case 'r':
                return '\r';
            case 't':
                return '\t';
            default:
                return c;
        }
    }

    private static final class TokenResponse {
        final String accessToken;
        final String tokenType;
        final long expiresIn;
        final String scope;

        TokenResponse(String accessToken, String tokenType, long expiresIn, String scope) {
            this.accessToken = accessToken;
            this.tokenType = tokenType;
            this.expiresIn = expiresIn;
            this.scope = scope;
        }
    }
}
